packages/flutter/lib/src/widgets/web_image.dart

```diff
@@ -0,0 +1,184 @@
+          loadingBuilder: loadingBuilder,
+          frameBuilder: frameBuilder,
+          errorBuilder: errorBuilder,
+        );
+
+  /// The resource which provides the image.
+  final WebImageProvider provider;
+
+  /// {@macro flutter.widgets.Image.loadingBuilder}
+  final ImageLoadingBuilder? loadingBuilder;
+
+  /// {@macro flutter.widgets.Image.frameBuilder}
+  final ImageFrameBuilder? frameBuilder;
+
+  /// {@macro flutter.widgets.Image.errorBuilder}
+  final ImageErrorWidgetBuilder? errorBuilder;
+
+  @override
+  State<WebImage> createState() => _WebImageState();
+}
+
+class _WebImageState extends State<WebImage> {
+  bool checkedIfBytesCanBeFetched = false;
+  bool imageBytesCanBeFetched = true;
+  @override
+  void initState() {
+    super.initState();
+    widget.provider.checkIfImageBytesCanBeFetched().then((bool canBeFetched) {
+      // We may have become unmounted in the time it took to check if the bytes
+      // could be fetched.
+      if (mounted) {
+        setState(() {
+          checkedIfBytesCanBeFetched = true;
+          imageBytesCanBeFetched = canBeFetched;
+        });
+      }
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    if (!checkedIfBytesCanBeFetched) {
+      return widget.loadingBuilder
+              ?.call(context, const SizedBox.shrink(), null) ??
+          const SizedBox.shrink();
+    }
+
+    if (imageBytesCanBeFetched) {
+      return Image(
+        image: widget.provider._networkImage!,
+        key: widget.key,
+        loadingBuilder: widget.loadingBuilder,
+        frameBuilder: widget.frameBuilder,
+        errorBuilder: widget.errorBuilder,
+      );
+    } else {
+      return impl.createImgElementWidget(
+        widget.provider._imgElementProvider!,
+        key: widget.key,
+        loadingBuilder: widget.loadingBuilder,
+        frameBuilder: widget.frameBuilder,
+        errorBuilder: widget.errorBuilder,
+      );
+    }
+  }
+}
+
+Future<void> precacheWebImage(
+  WebImageProvider provider,
+  BuildContext context, {
+  Size? size,
+  ImageErrorListener? onError,
+}) {
+  return provider.checkIfImageBytesCanBeFetched().then((bool canBeFetched) {
+    if (canBeFetched) {
+      if (context.mounted) {
+        return precacheImage(
+          provider._networkImage!,
+          context,
+          size: size,
+          onError: onError,
+        );
+      }
+    }
+
+    if (context.mounted) {
+      return impl.precacheImgElement(
+        provider._imgElementProvider!,
+        onError: onError,
+      );
+    }
+  });
+}
+
+class WebImageProvider {
+  WebImageProvider(this.src);
+
+  final String src;
+
+  bool? _imageBytesCanBeFetched;
+
+  NetworkImage? _networkImage;
+
+  ImgElementProvider? _imgElementProvider;
+
+  Future<bool> checkIfImageBytesCanBeFetched() {
+    if (_imageBytesCanBeFetched != null) {
+      return SynchronousFuture<bool>(_imageBytesCanBeFetched!);
+    }
+    return impl.checkIfImageBytesCanBeFetched(src).then((bool result) {
```

Comment:
@harryterkelsen 
I should be possible to eliminate CORS automatic check, and just specify that I know it will always be CORS.
I.e. "bool? forcedCors", something like that:
```dart
class WebImageProvider {
  WebImageProvider(this.src, {this.cssFilter, this.forcedCors});

  final String src;

  final String? cssFilter;

  final bool? forcedCors;

  bool? _imageBytesCanBeFetched;

  NetworkImage? _networkImage;

  ImgElementProvider? _imgElementProvider;

  Future<bool> checkIfImageBytesCanBeFetched() {
    if (_imageBytesCanBeFetched != null) {
      return SynchronousFuture<bool>(_imageBytesCanBeFetched!);
    }

    if (forcedCors == true) {
      _initImageBytesCanBeFetched(false);
      return SynchronousFuture<bool>(false);
    }

    return impl.checkIfImageBytesCanBeFetched(src).then((bool result) {
      _initImageBytesCanBeFetched(result);
      return result;
    });
  }

  void _initImageBytesCanBeFetched(bool value) {
    _imageBytesCanBeFetched = value;
    if (_imageBytesCanBeFetched!) {
      _networkImage = NetworkImage(src);
    } else {
      _imgElementProvider = impl.createImgElementProvider(src, cssFilter: cssFilter);
    }
  }
}
```